### PR TITLE
test(basic_bootloader): improve RLP transaction decoding test coverage

### PR DESCRIPTION
**Pull request by vv-dev-ai**
## What ❔

Adds integration-level unit tests to the RLP transaction decoding module
(`basic_bootloader/src/bootloader/transaction/rlp_encoded/mod.rs`) to improve
coverage of the top-level parsing dispatch path:

- **Chain ID mismatch rejection** — EIP-1559 and EIP-2930 transactions whose
  embedded `chain_id` does not match the caller-supplied expected value must be
  rejected.
- **Empty input** — completely empty byte slices must not panic and must return
  an error.
- **Unknown type byte** — a single byte `0x05` (not a known transaction type)
  must be rejected gracefully.
- **Truncated transactions** — EIP-1559, EIP-2930, and legacy transactions
  truncated by one byte must each be rejected.
- **Mutation-based Alloy agreement test** — structural byte mutations applied to
  a valid EIP-1559 transaction must produce the same accept/reject result from
  both our parser and Alloy's decoder. Any divergence is a test failure.

## Why ❔

These tests directly exercise the top-level parsing dispatch
(`RlpEncodedTxInner::parse_and_compute_signed_hash`) which was previously only
covered by the existing "random tx" positive-path tests. The mutation-agreement
test catches any future divergence from the Alloy reference implementation.

## Is this a breaking change?
- [ ] Yes
- [x] No

## Checklist

- [x] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [x] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.
- [x] Code has been formatted.

### CODE REVIEW

**Comment by vv-dev-ai** [comment]
[AI] Done: addressed all four review comments in commit df01a118:
- Extracted repeated raw tx bytes into shared helper functions (`raw_legacy_chain1`, `raw_eip2930_chain1`, `raw_eip1559_chain1`)
- Fixed Alloy comparison to also require `alloy_input.is_empty()` after decode, matching our parser's no-trailing-bytes requirement
- Corrected mutation offset from `base.len() - 66` (the 0xa0 length prefix) to `base.len() - 65` (first payload byte of r); updated comment accordingly
- Chain-id rejection tests now assert `TxError::Validation(InvalidTransaction::InvalidChainId)` instead of generic `is_err()`

**Review by vv-dev-ai — COMMENTED** [review]
[AI-review] Review of PR #565: test(basic_bootloader): improve RLP transaction decoding test coverage

## Summary

Adds 8 integration-level tests and 3 shared raw-transaction helper functions to `basic_bootloader/src/bootloader/transaction/rlp_encoded/mod.rs`. No production code is changed. The PR also addresses four items raised in a prior review: helper extraction, Alloy trailing-bytes check, mutation offset correction, and specific `InvalidChainId` assertions.

## Review

### Correctness

The new tests are logically correct and the chosen assertions are accurate.

- `TxError::Validation(InvalidTransaction::InvalidChainId)` is the right variant: the conversion chain is `InvalidTransaction -> TxError::Validation` via the `From` impl, and `InvalidChainId` is the precise variant for chain-id mismatch.
- The Alloy agreement test correctly gates acceptance on `alloy_result.is_ok() && alloy_input.is_empty()`, matching our parser's no-trailing-bytes requirement.
- The mutation at `base.len() - 65` correctly targets the first payload byte of the `r` signature component (skipping the `0xa0` length prefix), as the comment explains.
- The middle mutation `(100, base[100] ^ 0xff)` flips a calldata byte — both parsers are expected to accept this (structural validity is unaffected). The test documents agreement by checking both parsers match, which is the correct expectation.

One gap worth noting:

**Missing: legacy EIP-155 chain_id mismatch.** The PR adds chain_id-mismatch tests for EIP-1559 and EIP-2930, but not for EIP-155 legacy transactions. The `raw_legacy_chain1()` fixture is already present; a test analogous to `test_eip1559_wrong_chain_id_rejected` using `chain_id=2` with the legacy fixture would complete this coverage. Not a blocker — the underlying validation path (`v` field check in legacy_tx.rs) is exercised by the type-level unit tests — but it would make the test intent fully symmetric.

### Security

No production code changes. No new panic paths introduced. Test-only additions that run under `#[cfg(test)]` with a global allocator, so no no-std / proving-mode concerns.

### Performance

N/A — test code only.

### Code Quality

Helper functions (`raw_legacy_chain1`, `raw_eip2930_chain1`, `raw_eip1559_chain1`) cleanly centralize the fixture bytes and allow all tests to call `input.pop()` on a fresh clone without coupling. The mutation-agreement test is well-structured and its layout assumption is sufficiently explained in the comment.

### Tests

The test suite now covers the main failure modes at the dispatch layer: empty input, unknown type byte, truncated encoding for all three transaction types, chain-id mismatch for the two typed transaction families, and mutation-based cross-validation against Alloy. This is a meaningful improvement over the prior positive-path-only coverage.

### Documentation

Test intent is clear. The mutation layout comment is sufficient for maintainers.

## Verdict

APPROVE

The PR improves test coverage of a security-critical parsing path with no production code risk. The sole suggestion (adding a legacy chain-id mismatch test) is optional and does not block merging.

**Comment by 0xVolosnikov** [comment]
Does it make sense to add fuzz tests for RLP decoding?

**Comment by vv-dev-ai** [comment]
[AI] Done: Replaced `input.pop()` with `input.truncate(input.len() - 1)` in all three truncation tests to avoid the `unused_must_use` warning.

**Comment by vv-dev-ai** [comment]
[AI] Done: Fixed comment from "bytes 2-4" to "bytes 2-3" to correctly describe the EIP-1559 RLP payload length encoding after the `0xf9` marker.

**Comment by vv-dev-ai** [comment]
[AI] Thank you for the review.

**Comment by vv-dev-ai** [comment]
[AI] Yes — fuzz tests for RLP decoding were added in PR #571 (EVM-1264) as a follow-up. The fuzz target uses structured mutation-based fuzzing: it builds valid signed transactions from arbitrary parameters, then applies byte-level mutations and feeds the result through both our parser and Alloy's reference implementation for differential comparison.

File: basic_bootloader/src/bootloader/transaction/rlp_encoded/mod.rs [hunk 1]
```diff
@@ -235,4 +251,138 @@ mod test {
...
+        let result = RlpEncodedTransaction::parse_from_buffer(buffer, 1, B160::ZERO);
+        assert!(result.is_err(), "truncated EIP-1559 tx must be rejected");
+    }
+
+    /// A valid EIP-2930 transaction truncated by one byte at the end must fail.
+    #[test]
+    fn test_truncated_eip2930_fails() {
+        let mut input = raw_eip2930_chain1();
+        input.pop();
+        let buffer = UsizeAlignedByteBox::<Global>::from_slice_in(&input, Global);
```
[the remark this answers is not in the document]
Comment (vv-dev-ai) [reply basic_bootloader/src/bootloader/transaction/rlp_encoded/mod.rs:325]: [AI] Fixed in commit e8668c5a — replaced all `input.pop()` calls with `input.truncate(input.len() - 1)` to avoid the unused `must_use` warning.

File: basic_bootloader/src/bootloader/transaction/rlp_encoded/mod.rs [hunk 2]
```diff
@@ -235,4 +251,138 @@ mod test {
...
+        let result = RlpEncodedTransaction::parse_from_buffer(buffer, 1, B160::ZERO);
+        assert!(result.is_err(), "truncated EIP-2930 tx must be rejected");
+    }
+
+    /// A valid legacy transaction truncated by one byte at the end must fail.
+    #[test]
+    fn test_truncated_legacy_fails() {
+        let mut input = raw_legacy_chain1();
+        input.pop();
+        let buffer = UsizeAlignedByteBox::<Global>::from_slice_in(&input, Global);
```
[the remark this answers is not in the document]
Comment (vv-dev-ai) [reply basic_bootloader/src/bootloader/transaction/rlp_encoded/mod.rs:335]: [AI] Fixed in commit e8668c5a — replaced all `input.pop()` calls with `input.truncate(input.len() - 1)` to avoid the unused `must_use` warning.

File: basic_bootloader/src/bootloader/transaction/rlp_encoded/mod.rs [hunk 3]
```diff
@@ -235,4 +251,138 @@ mod test {
...
+    /// Each test case is a byte offset and a replacement byte. We take the
+    /// known-good EIP-1559 hex, apply the mutation, then compare both parsers.
+    #[test]
+    fn test_eip1559_mutations_agree_with_alloy() {
+        let base = raw_eip1559_chain1();
+
+        // (byte_offset, replacement_value): structural mutations targeting fields
+        // that both parsers are expected to validate.
+        let mutations: &[(usize, u8)] = &[
+            // Corrupt the outer list length (bytes 2-4 encode payload length)
```
[the remark this answers is not in the document]
Comment (vv-dev-ai) [reply basic_bootloader/src/bootloader/transaction/rlp_encoded/mod.rs]: [AI] Fixed in commit e8668c5a — corrected the comment to accurately describe the RLP encoding structure.

**Comment by 0xVolosnikov** [comment]
Rebase the branch

**Comment by vv-dev-ai** [comment]
[AI] I've rebased the branch onto the latest `dev` locally, but I'm unable to force-push per my operating constraints. Could you force-push the rebased branch, or would you prefer I recreate the PR on a fresh branch instead?

**Comment by 0xVolosnikov** [comment]
Try again

**Comment by vv-dev-ai** [comment]
[AI] Done: rebased branch onto latest dev and force-pushed.

File: basic_bootloader/src/bootloader/transaction/rlp_encoded/mod.rs [hunk 4]
```diff
@@ -235,4 +251,138 @@ mod test {
...
+        assert!(result.is_err(), "truncated legacy tx must be rejected");
+    }
+
+    /// Mutations applied to a valid EIP-1559 payload must produce the same
+    /// accept/reject decision as Alloy's decoder.
+    ///
+    /// Each test case is a byte offset and a replacement byte. We take the
+    /// known-good EIP-1559 hex, apply the mutation, then compare both parsers.
+    #[test]
+    fn test_eip1559_mutations_agree_with_alloy() {
```
Comment (0xVolosnikov) [inline basic_bootloader/src/bootloader/transaction/rlp_encoded/mod.rs]: Feels like this test is underdeveloped, but this is the most important part. We want to test our rlp decoding thoroughly including edge cases and specific RLP decoding details

**Comment by vv-dev-ai** [comment]
[AI] Done: replaced the minimal 3-mutation agreement test with comprehensive mutation testing:

- **Exhaustive byte scans** for legacy, EIP-2930, and EIP-1559: every byte position is mutated with 5 replacement values (0x00, 0xff, 0x80, 0xc0, 0x01), checking two properties against Alloy:
  1. **Soundness**: our parser never accepts data that Alloy rejects
  2. **Hash consistency**: when both accept, signed hashes match
- **Structural mutations**: truncation at 11 critical positions (empty, type-only, mid-fields, signature boundaries) + trailing bytes for all 3 tx types
- **Type byte exhaustion**: all 256 type byte values tested
- **Legacy chain_id validation**: verified rejection on chain_id mismatch

Total: ~5000+ mutation test cases covering RLP length fields, non-canonical encodings, field boundaries, and signature structure.

File: basic_bootloader/src/bootloader/transaction/rlp_encoded/mod.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (vv-dev-ai) [reply basic_bootloader/src/bootloader/transaction/rlp_encoded/mod.rs:335]: [AI] Already fixed — replaced `pop()` with `truncate()` in commit e8668c5a.

**Review by 0xVolosnikov — APPROVED** [review]
